File: 518-coin-change-ii/coin-change-ii.cpp
#define ull unsigned long long
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        int n = coins.size();
        vector<vector<ull>> dp(n, vector<ull>(amount + 1, 0));

        for (int i = 0; i < n; i++) dp[i][0] = 1;
        
        for (int j = 1; j <= amount; j++) {
            if (j % coins[0] == 0) dp[0][j] = 1;
        }

        for (int i = 1; i < n; i++) {
            for (int j = 0; j <= amount; j++) {
                ull not_take = dp[i - 1][j];
                ull take = 0;
                if (j >= coins[i]) {
                    take = dp[i][j - coins[i]];
                }
                dp[i][j] = take + not_take;
            }
        }

        return dp[n - 1][amount];
    }
};

File: 518-coin-change-ii/coin-change-ii.go
func solve(coins []int, amount int, idx int, dp [][]int) int {
    if idx < 0 || amount < 0 { return 0 }
    if amount == 0 { return 1 }

    if dp[idx][amount] != -1 {
        return dp[idx][amount]
    }

    take, not_take := 0, 0
    if amount >= coins[idx] {
        take = solve(coins, amount - coins[idx], idx, dp)
    }
    not_take = solve(coins, amount, idx-1, dp)

    dp[idx][amount] = take + not_take
    return dp[idx][amount]
}

func change(amount int, coins []int) int {
    n := len(coins)
    dp := make([][]int, n)
    for i := range dp {
        dp[i] = make([]int, amount + 1)
        for j := range dp[i] {
            dp[i][j] = -1
        }
    }

    ways := solve(coins, amount, n-1, dp)

    return ways
}